Add working dockerfile and update readme

ctkcoding opened the pull request.

ctkcoding commented:
Built the referenced docker container and pushed to https://hub.docker.com/repository/docker/ctkcoding/dir2cast

ctkcoding commented:
@ben-xo not seeing the option to request a review like I'm used to, so tagging you in the comments here instead

ben-xo commented:
Hey, thanks for this. 

I have some feedback, which I'll leave as a review. However it's worth noting that dir2cast already has a docker example (using nginx and php-fpm) - checkout the `docker-compose.yml` file already in the repo. I guess it doesn't really work the same way, and there are pros and cons - and personal preferences - to both approaches. Also I guess the  current example isn't actually documented in the README.

e.g do you want dir2cast and config on your filesystem or hidden within the container? are you more of an Apache guy or a Nginx fan? (i'm the latter, but I am not a php-fpm fan so it kind of cuts both ways).

Anyway i'll leave you a proper review next.

ben-xo commented on Dockerfile:
```diff
@@ -0,0 +1,35 @@
+    TITLE= \
+    AUTO_SAVE_COVER_ART=
+
+# add any dir2cast.ini options as env vars here
+# todo - making this reusable for all env vars
+RUN echo '; review options in dir2cast.ini source' > /var/www/html/dir2cast.ini
+RUN echo "ITEM_COUNT = ${ITEM_COUNT}" >> /var/www/html/dir2cast.ini
+RUN echo "MP3_DIR = ${MP3_DIR}" >> /var/www/html/dir2cast.ini
+RUN if [[ $TITLE ]]; then echo "TITLE = ${TITLE}" >> /var/www/html/dir2cast.ini; fi;
+RUN if [[ $MP3_URL ]]; then echo "MP3_URL = ${MP3_URL}" >> /var/www/html/dir2cast.ini; fi;
+
+RUN if [[ $IMAGE ]]; then echo "IMAGE = ${IMAGE}" >> /var/www/html/dir2cast.ini; fi;
+RUN if [[ $ITUNES_EXPLICIT ]]; then echo "ITUNES_EXPLICIT = ${ITUNES_EXPLICIT}" >> /var/www/html/dir2cast.ini; fi;
+RUN if [[ $DESCRIPTION ]]; then echo "DESCRIPTION = ${DESCRIPTION}" >> /var/www/html/dir2cast.ini; fi;
+RUN if [[ $ITUNES_AUTHOR ]]; then echo "ITUNES_AUTHOR = ${ITUNES_AUTHOR}" >> /var/www/html/dir2cast.ini; fi;
+RUN if [[ $RECURSIVE_DIRECTORY_ITERATOR ]]; then echo "RECURSIVE_DIRECTORY_ITERATOR = ${RECURSIVE_DIRECTORY_ITERATOR}" >> /var/www/html/dir2cast.ini; fi;
+RUN if [[ $AUTO_SAVE_COVER_ART ]]; then echo "AUTO_SAVE_COVER_ART = ${AUTO_SAVE_COVER_ART}" >> /var/www/html/dir2cast.ini; fi;
```
so, I assume what you intended was for someone to be able to do e.g. `docker run  -d -p 8080:80 -v podcast_volume:/var/www/html/episodes -e IMAGE=... -e ITUNES_EXPLICIT=... ctkcoding/dir2cast:v0.1` etc etc. but, these `RUN` lines run at *build* time, not run time. 

There are two standard approaches to deal with this.

1. the user should edit dir2cast.ini and then mount it at run time with `-v dir2cast.ini:/var/www/html/dir2cast.ini`
2. use `envsubst` to replace placeholders in the baked in dir2cast.ini at run time, so you can do `-e THING=whatever` like i imagine you intended. (You will need to set a default for every single one i think). This way is harder but nicer for the user

ctkcoding replied:
Updated to support ini creation through environment variables at container run time rather than image build time

ben-xo commented on dir2cast.php:
```diff
@@ -1246,7 +1246,13 @@ protected function scan()
             {
                 http_response_code(404);
-                throw new Exception("No content yet.");
+                if (0 == $item_count) {
+                    throw new Exception("Item count zero.");
+                }
+                if (self::$EMPTY_PODCAST_IS_ERROR) {
+                    throw new Exception("Self empty podcast error.");
+                }
```
because the outer condition is `if(self::$EMPTY_PODCAST_IS_ERROR && 0 == $item_count)`, this will always throw "Item count is zero". The "Self empty podcast error" exception is unreachable.

ctkcoding replied:
Can't track down the branch, but I had pulled this in from another branch on the main project, just like the attribution changes in the readme. Reverting since I don't see this on the main branch

ben-xo commented on dir2cast.php:
```diff
@@ -1,7 +1,7 @@
 
 /******************************************************************************
- * Copyright (c) 2008-2022, Ben XO (me@ben-xo.com).
+ * Copyright (c) 2008-2023, Ben XO (@benxo on most platforms).
```
thanks for this.

ben-xo commented on Dockerfile:
```diff
@@ -0,0 +1,35 @@
+# copy source files to docker 
+COPY ./dir2cast.php /var/www/html/
+# skip creating ini. generated above
+# COPY ./dir2cast.ini /var/www/html/
+COPY ./getID3/ /var/www/html/getID3/
```
so, the code for getID3 and also the dir2cast.ini file will be downloadable, unless you add some .htaccess to block them. There's instructions for that in the README.

ctkcoding replied:
This type of server isn't in my expertise so I might be missing something, but it looks like the ini file has to be in the same relative position to the php file the way things are now so it seems like admins would need to set up .htaccess to block whether its running from my docker image or any of the ways listed in the readme

ctkcoding replied:
Updated with .htaccess to block any .php or .ini requests, and seeing a 403 when I try /getID3/ as the path.

ben-xo reviewed (commented):
Some things I spotted.

ctkcoding commented:
Thanks for the feedback - will clean up & address over the weekend! Also re why I'm not into docker compose - I run all my docker apps using images from docker hub and it's more convenient accessing it through a gui on my laptop than sshing in to the server itself to manage things. Re apache vs nginx - my real preference is dealing with a single port on my host machine because it's all abstracted by docker

ben-xo commented:
> Thanks for the feedback - will clean up & address over the weekend! Also re why I'm not into docker compose - I run all my docker apps using images from docker hub and it's more convenient accessing it through a gui on my laptop than sshing in to the server itself to manage things. Re apache vs nginx - my real preference is dealing with a single port on my host machine because it's all abstracted by docker

docker-compose is supported by both OrbStack and the official Docker desktop app, as far as i know!

ctkcoding commented:
Updated to resolve issues
- run-time environment variables that generate new dir2cast.ini file
- reverted php code change erroneously pulled from other branch
- blocked getID3 and dir2cast.ini access from path manipulation

ctkcoding commented:
@ben-xo any comments/feedback on the latest version or does this look good to merge?

ctkcoding commented:
@ben-xo ?
